File: src/python/aoc2.py
import re


def answer1(all_rounds):
    # A|X: Rock
    # B|Y: Paper
    # C|Z: Scissors
    return (
        6 * len(re.findall(r'A Y|B Z|C X', all_rounds))  # Win
        + 3 * len(re.findall(r'A X|B Y|C Z', all_rounds))  # Draw
        + 1 * all_rounds.count('X')  # Rock
        + 2 * all_rounds.count('Y')  # Paper
        + 3 * all_rounds.count('Z')  # Scissors
    )


def answer2(all_rounds):
    # A: Rock
    # B: Paper
    # C: Scissors
    # X: Lose
    # Y: Draw
    # Z: Win
    return (
        6 * all_rounds.count('Z')  # Win
        + 3 * all_rounds.count('Y')  # Draw
        + 1 * len(re.findall(r'A Y|B X|C Z', all_rounds))  # Rock
        + 2 * len(re.findall(r'A Z|B Y|C X', all_rounds))  # Paper
        + 3 * len(re.findall(r'A X|B Z|C Y', all_rounds))  # Scissors
    )


def main():
    all_rounds = load()
    print(f'Answer 1: {answer1(all_rounds)}')
    print(f'Answer 2: {answer2(all_rounds)}')


def load():
    with open('input/aoc2', encoding='utf-8') as file_:
        return file_.read()


def test_answer1():
    assert answer1(load()) == 11841


def test_answer2():
    assert answer2(load()) == 13022


main()

File: src/rust/aoc2.rs
pub use crate::loaders::file_to_lines as load;
pub const DATA: &str = "input/aoc2";

const WIN: u32 = 6;
const DRAW: u32 = 3;

const LOSE: u32 = 0;
const ROCK: u32 = 1;
const PAPER: u32 = 2;
const SCISSORS: u32 = 3;

pub fn answer1<I>(rounds: I) -> u32
where
    I: Iterator,
    I::Item: AsRef<str>,
{
    rounds
        .map(|round| match round.as_ref() {
            "A X" => ROCK + DRAW,
            "B X" => ROCK + LOSE,
            "C X" => ROCK + WIN,
            "A Y" => PAPER + WIN,
            "B Y" => PAPER + DRAW,
            "C Y" => PAPER + LOSE,
            "A Z" => SCISSORS + LOSE,
            "B Z" => SCISSORS + WIN,
            "C Z" => SCISSORS + DRAW,
            _ => unreachable!("All game cases should be covered"),
        })
        .sum()
}

pub fn answer2<I>(rounds: I) -> u32
where
    I: Iterator,
    I::Item: AsRef<str>,
{
    rounds
        .map(|round| match round.as_ref() {
            "A X" => LOSE + SCISSORS,
            "B X" => LOSE + ROCK,
            "C X" => LOSE + PAPER,
            "A Y" => DRAW + ROCK,
            "B Y" => DRAW + PAPER,
            "C Y" => DRAW + SCISSORS,
            "A Z" => WIN + PAPER,
            "B Z" => WIN + SCISSORS,
            "C Z" => WIN + ROCK,
            _ => unreachable!("All game cases should be covered"),
        })
        .sum()
}

#[cfg(test)]
mod tests {
    use super::*;

    const MOCK_DATA: &str = "A Y\nB X\nC Z\n";

    #[test]
    fn test_answer1_mock_data() {
        assert_eq!(answer1(MOCK_DATA.lines()), 15)
    }

    #[test]
    fn test_answer2_mock_data() {
        assert_eq!(answer2(MOCK_DATA.lines()), 12)
    }

    #[test]
    fn test_answer1() {
        assert_eq!(answer1(load(DATA)), 11841)
    }

    #[test]
    fn test_answer2() {
        assert_eq!(answer2(load(DATA)), 13022)
    }
}
